### ISSUE TITLE
Add way to exclude file from watch but not for transpiling

### ISSUE THREAD

**Issue by arthurfiorette** [beenotung/live-tsc#3]
my postHook code generates some files inside src, to prevent an infinite loop, i add all of them in the `--exclude` argument. But they also gets excluded from being transpiled to target folder.

**Comment by beenotung** [comment]
A workaround is to add one more post hook to transpile the generated files.

The post hooks are ran sequentially.

And the server is restarted after all post hooks complete.

**Comment by arthurfiorette** [comment]
Should work... I'll try as soon as I arrive home.

another cool feature that could be integrated as well is set optional post hooks, that only runs when some files have changed, like `--post-hook ./directory-to-listen=command to only run when directory changed`

**Comment by beenotung** [comment]
Do you prefer having conditional post hook, or having multiple instances of live-tsc in watch mode, each having different src,dest,post-hook?

**Comment by arthurfiorette** [comment]
Actually, just one live tsc instance would be awesome, as the "ohter" instance would just to serve the post hook

**Comment by beenotung** [comment]
You want to generate some files into the src directory, maybe you're looking for pre-hook instead of post-hook?

Should the generated files be transpiled?

**Comment by beenotung** [comment]
I've setup an example [1] that involve code generation.

In this example, you can start the `live-tsc` by running `npm run dev`. After that, changing the content of `src/enums.macro.ts` will cause `live-tsc` to restart the server for several rounds.

It eventually stop looping, probably due to the [change detection on transpiled code](https://github.com/beenotung/live-tsc/blob/b2fbfba/src/core.ts#L368). But the several restarts is definitely sub-optimal.

My idea is to add `pre-hook` to support logics like code generation, and changes detected when running pre-hooks will be buffered and deduplicated.

[1] https://github.com/beenotung/live-tsc/blob/exp-loop/examples/package.json

**Comment by arthurfiorette** [comment]
> Should the generated files be transpiled?

Yes

**Comment by arthurfiorette** [comment]
A pre hook would be awesome, but as I said before, you could add a way to only run pre/post when some files were changed.

**Comment by beenotung** [comment]
> Should work... I'll try as soon as I arrive home.
> 
> another cool feature that could be integrated as well is set optional post hooks, that only runs when some files have changed, like `--post-hook ./directory-to-listen=command to only run when directory changed`

I agree this suggestion in general, still thinking a syntax to specify the paths to watch.

Using equal sign as deliminator will break environment variable assignment, e.g. `--post-hook "MODE=quick npm test"` should run the command `npm test` with env var `MODE` set as `quick`

**Comment by beenotung** [comment]
You can now only run post hook when a specific file (or files) changes with `--post-hook command#watch:file1,file2`.

Also added new option `--post-script name` as shortcut of `--post-hook "npm run name"`. This option also support specifying watch files. 

This patch is released as v1.5.0 on npm :tada: 

Will add pre hook next day.

**Comment by arthurfiorette** [comment]
Awesome! Will try it tomorrow.

Does #watch: support patterns, like: `#watch:./src/*.{ts,js}`?

> Using equal sign as deliminator will break environment variable assignment

What about escaping it? `=` being used to files, but `\=` is kept in the command. Because some time, don't know when, someone will try to use `#watch` inside his command and it will break :)

**Comment by arthurfiorette** [comment]
> Also added new option --post-script name as shortcut of --post-hook "npm run name"

Will you implement `--pre-hook` and `--pre-script` too?

**Comment by beenotung** [comment]
> Awesome! Will try it tomorrow.
> 
> Does #watch: support patterns, like: `#watch:./src/*.{ts,js}`?
> 
> > Using equal sign as deliminator will break environment variable assignment
> 
> What about escaping it? `=` being used to files, but `\=` is kept in the command. Because some time, don't know when, someone will try to use `#watch` inside his command and it will break :)

Do you have a relative large amount of files to watch? I can make it support watching directory as well. But don't quite want to support glob expansion to keep it simple.

Another syntax for watch is `--post-hook <command>#live-tsc-watch:<files>`. This format has less chance to conflict with user command. But I suppose `#` is rarely used because it means comment in bash command.

A workaround to avoid conflict is to the put the command into a npm script, then call it with `--post-script <script-name>#watch:<files>`.

**Comment by beenotung** [comment]
> > Also added new option --post-script name as shortcut of --post-hook "npm run name"
> 
> Will you implement `--pre-hook` and `--pre-script` too?

I will add pre hook and pre script later today, it will skip changes detected during pre hook execution.

**Comment by arthurfiorette** [comment]
> Do you have a relative large amount of files to watch? I can make it support watching directory as well. But don't quite want to support glob expansion to keep it simple.

I could use to listen to a directory, that's fine.

**Comment by arthurfiorette** [comment]
> Another syntax for watch is --post-hook <command>#live-tsc-watch:<files>. This format has less chance to conflict with user command. But I suppose # is rarely used because it means comment in bash command.

See #4

**Comment by arthurfiorette** [comment]
Also, regarding the usage of `fs.watch`, some events were being fired twice. Started looking out and noticed this:  https://stackoverflow.com/a/33047844/14681561.

I think chokidar should be used instead too.

**Comment by arthurfiorette** [comment]
@beenotung any updates?

**Comment by beenotung** [comment]
I think we already address this issue with the `#watch:files` syntax?

(I will test if changing non-listed files will still trigger the hook later)

I want to avoid chokidar if the issue(s) can be solved here directly to keep this package lightweight.

If you want to discuss other features (e.g. glob) we can discuss in separate issue.

**Comment by beenotung** [comment]
One of the issues of fs.watch mentioned in the linked stackoverflow post is some editor will trigger multiple change events when saving a file.

This situation is handled by skipping further execution of the hook if it's already running.

https://github.com/beenotung/live-tsc/commit/859dc8980fb522a4be973cc2217369da4fdc2c1e
